validator/sawtooth_validator/networking/dispatch.py

```diff
@@ -104,53 +103,59 @@ def add_handler(self, message_type, handler, executor):
                 _HandlerManager(executor, handler))
 
     def _process(self, message_id):
-        with self._condition:
-            _, connection_id, \
-                message, collection = self._message_information[message_id]
+        _, connection_id, \
+            message, collection = self._message_information[message_id]
         try:
             handler_manager = next(collection)
             future = handler_manager.execute(connection_id, message.content)
             future.add_done_callback(partial(self._determine_next, message_id))
         except IndexError:
             # IndexError is raised if done with handlers
-            with self._condition:
-                del self._message_information[message_id]
+            del self._message_information[message_id]
 
     def _determine_next(self, message_id, future):
         if future.result().status == HandlerStatus.DROP:
-            with self._condition:
-                del self._message_information[message_id]
+            del self._message_information[message_id]
 
         elif future.result().status == HandlerStatus.PASS:
             self._process(message_id)
 
         elif future.result().status == HandlerStatus.RETURN_AND_PASS:
-            with self._condition:
-                connection, connection_id, \
-                    original_message, _ = self._message_information[message_id]
+            connection, connection_id, \
+                original_message, _ = self._message_information[message_id]
 
             message = validator_pb2.Message(
                 content=future.result().message_out.SerializeToString(),
                 correlation_id=original_message.correlation_id,
                 message_type=future.result().message_type)
-
-            self._send_message[connection](msg=message,
-                                           connection_id=connection_id)
+            try:
+                self._send_message[connection](msg=message,
+                                               connection_id=connection_id)
+            except KeyError:
+                LOGGER.info("Can't send message %s back to "
+                            "%s because connection %s not in dispatcher",
+                            get_enum_name(message.message_type), connection_id,
+                            connection)
             self._process(message_id)
 
         elif future.result().status == HandlerStatus.RETURN:
-            with self._condition:
-                connection, connection_id,  \
-                    original_message, _ = self._message_information[message_id]
+            connection, connection_id,  \
+                original_message, _ = self._message_information[message_id]
```

Comment:
I think the .get plus del in these lines don't need to be atomic, since it would be fine for several .gets to happen across threads and then the dels. For any given message (tied to a message_id), the _determine_next function will be sequential.